File: Demo/MNG-Ads-SDK/InfeedViewController.m
//
//  ParalaxViewController.m
//  MNG-Ads-SDK-Demo
//
//  Created by HtrimechMac on 13/11/2019.
//  Copyright © 2019 Bensalah Med Amine. All rights reserved.
//

#import "InfeedViewController.h"

#define infeeRowPosition 8

@interface InfeedViewController ()<MNGAdsAdapterInfeedDelegate>{
    MNGAdsSDKFactory *infeedAdsFactory;
    UIView *_infeedView;
    BOOL checkTableView;
}

@end

@implementation InfeedViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    [[UIDevice currentDevice] beginGeneratingDeviceOrientationNotifications];
    [[NSNotificationCenter defaultCenter]
     addObserver:self selector:@selector(orientationChanged:)
     name:UIDeviceOrientationDidChangeNotification
     object:[UIDevice currentDevice]];
}

-(void)viewDidAppear:(BOOL)animated {
    
    [self loadTableView:nil];
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void) orientationChanged:(NSNotification *)note
{
    [self.tableView reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:infeeRowPosition inSection:0]] withRowAnimation:UITableViewRowAnimationAutomatic];
    
}

-(void)dealloc{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    [infeedAdsFactory releaseMemory];
    infeedAdsFactory   = nil;
    
    
}
- (IBAction)openMenuView:(id)sender {
    [[APP_DELEGATE drawerViewController]openMenu];
}
- (void)loadInfeedView {
    if (_infeedView) {
        [_infeedView removeFromSuperview];
        _infeedView = nil;
    }
    
    if (!checkTableView) {
        [self.tableView reloadData];
    }
   
    infeedAdsFactory = [[MNGAdsSDKFactory alloc]init];
    infeedAdsFactory.infeedDelegate = self;
    infeedAdsFactory.viewController = [APP_DELEGATE window].rootViewController;
    infeedAdsFactory.placementId = MNG_ADS_INFEED_PLACEMENT_ID;
    MNGPreference *preferences = [Utils getTestPreferences];
    
    MAdvertiseInfeedFrame * frameinfeed = [[MAdvertiseInfeedFrame alloc] initWithWidthDP:self.view.frame.size.width andInfeedRatio:INFEED_RATIO_16_9];
    [infeedAdsFactory loadInfeedInFrame:frameinfeed withPreferences:preferences];
    
}

- (IBAction)loadTableView:(id)sender {
  
    self.tableView.hidden = NO;
    self.infeedScrollView.hidden = YES;
    checkTableView = NO;
    [self loadInfeedView];
    
}

#pragma mark - Infeed Delegate
-(void)adsAdapter:(MNGAdsAdapter *)adsAdapter infeedDidLoad:(UIView *)adView preferredHeight:(CGFloat)preferredHeight{
    NSLog(@"adsAdapterInfeedDidLoad:preferredHeight: %f",preferredHeight);
    _infeedView = adView;
    _infeedView.frame = CGRectMake((self.view.bounds.size.width - adView.bounds.size.width )/2, 0, adView.bounds.size.width, adView.bounds.size.height);
    if (checkTableView) {
        if (_infeedView) {
            [_infeedSview addSubview:_infeedView];
        }

    } else {
        [self.tableView reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:infeeRowPosition inSection:0]] withRowAnimation:UITableViewRowAnimationAutomatic];
    }

}

-(void)adsAdapter:(MNGAdsAdapter *)adsAdapter infeedDidFailWithError:(NSError *)error{
    NSLog(@"%@",error);
    _infeedView = nil;
    if (!checkTableView) {
        [self.tableView reloadData];
    }
}
-(void)viewDidLayoutSubviews{
    
}
#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return 30;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.row == infeeRowPosition) {
        static NSString *CellIdentifier = @"paralax";
           UITableViewCell * cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];

//        = [self.tableView dequeueReusableCellWithIdentifier:CellIdentifier];

//           if (cell == nil)
//           {
//           }
    
        if (_infeedView) {
            [cell.contentView addSubview:_infeedView];
            [cell.contentView setClipsToBounds:YES];
            cell.backgroundColor = [UIColor clearColor];
            cell.contentView.backgroundColor = [UIColor clearColor];


        }
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        return cell;
    }else{
        if (indexPath.row == 0) {
            UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"1"];
            if (cell) {
                return cell;
            }
            cell = [[UITableViewCell alloc]initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"1"];
            cell.textLabel.text = @"Best practice Mngads : optimized use case for several ad formats on one page.";
            cell.textLabel.numberOfLines = 0;
            cell.textLabel.font = [UIFont systemFontOfSize:14];
            return cell;
        }
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"x"];
        if (cell) {
            return cell;
        }
        cell = [[UITableViewCell alloc]initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"x"];
        UIImageView *image = [[UIImageView alloc]initWithImage:[UIImage imageNamed:@"article"]];
        image.autoresizingMask = UIViewAutoresizingFlexibleHeight|UIViewAutoresizingFlexibleWidth;
        image.frame = cell.bounds;
        [cell addSubview:image];
        return cell;
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.row == infeeRowPosition) {
        if (_infeedView) {
           return self.view.frame.size.width * (3./5.) ;//_infeedView.frame.size.height;
        }
        return 0;
    }else{
        return 60;
    }
}

- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.row == infeeRowPosition) {
        if (_infeedView) {
            return self.view.frame.size.width * (3./5.) ;//_infeedView.frame.size.height;
        }
        return 0;
    }else{
        return 60;
    }
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.row == infeeRowPosition) {
        NSLog(@"index %ld",(long)indexPath.row);
    }else {
        NSLog(@"tableview seletected here %ld",(long)indexPath.row);
    }
}

- (IBAction)loadScrollView:(id)sender {
    self.tableView.hidden = YES;
    self.infeedScrollView.hidden = NO;
    checkTableView = YES;
    [self loadInfeedView];
    
}
/*
 #pragma mark - Navigation
 
 // In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
 // Get the new view controller using [segue destinationViewController].
 // Pass the selected object to the new view controller.
 }
 */

@end

File: DemoSwift/DemoSwift/Infeed/InfeedViewController.swift
//
//  InfeedViewController.swift
//  DemoSwift
//
//  Created by HtrimechMac on 07/06/2021.
//

import UIKit
import BlueStackSDK

class InfeedViewController:  UIViewController , UITableViewDelegate, UITableViewDataSource , MNGAdsSDKFactoryDelegate, MNGAdsAdapterInfeedDelegate   {
     
    
    // cell reuse id (cells that scroll out of view can be reused)
    let cellReuseIdentifier = "cell"
    let infeedRow = 6

    var infeedFactory: MNGAdsSDKFactory?
    var infeedView : UIView?
    var  myString = "cell"
    
    var index : Int = 0
    
    @IBOutlet weak var infeedTableView: UITableView!
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        // Register the table view cell class and its reuse id
        infeedTableView.register(UITableViewCell.self, forCellReuseIdentifier: cellReuseIdentifier)
        
        // (optional) include this line if you want to remove the extra empty cell divider lines
        // self.tableView.tableFooterView = UIView()
        
        // This view controller itself will provide the delegate methods and row data for the table view.
        infeedTableView.delegate = self
        infeedTableView.dataSource = self
    }
    override func viewDidAppear(_ animated: Bool) {
         super.viewDidAppear(animated)
         loadInfeed()

    }
    
    @IBAction func reloadInfeed(_ sender: Any) {
        loadInfeed()
    }
    // MARK: - Infeed
    func loadInfeed()  {
        if infeedView != nil {
            infeedView?.removeFromSuperview()
            infeedView = nil
            infeedTableView.reloadData()
            
        }
        infeedFactory = MNGAdsSDKFactory()
        infeedFactory?.placementId = "/\(MNG_ADS_APP_ID!)\( DemoSwiftConstants.PLACEMENTS.MNG_ADS_INFEED_PLACEMENT_ID)"
        infeedFactory?.infeedDelegate = self
        infeedFactory?.viewController = self
        let infeedFrame = MAdvertiseInfeedFrame(widthDP: self.view.frame.size.width, andInfeedRatio:INFEED_RATIO_16_9)
        infeedFactory?.loadInfeed(in: infeedFrame, withPreferences: nil)
    }
    
    
    func adsAdapter(_ adsAdapter: MNGAdsAdapter!, infeedDidLoad adView: UIView!) {
          print("infeed loaded")
          infeedView = adView
        infeedView!.frame = CGRect(x:(self.view.bounds.size.width - infeedView!.bounds.size.width )/2, y: 0, width: infeedView!.frame.size.width, height: infeedView!.frame.size.height)
        //refresh the indexPath
        let indexPath = IndexPath(row: infeedRow, section: 0)
        infeedTableView.reloadRows(at: [indexPath], with: .automatic)
    }
    
    func adsAdapter(_ adsAdapter: MNGAdsAdapter!, infeedDidFailWithError error: Error!) {
        print(error as Any)
        if infeedView != nil {
            infeedView?.removeFromSuperview()
            infeedView = nil
            infeedTableView.reloadData()
        }
            
    }
    
    
     // MARK: - Actions
    
    @IBAction func returnBack(_ sender: UIButton) {
        self.dismiss(animated: true, completion: nil)
    }

    // MARK: - UITableView Delegate
    
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 30
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        if indexPath.row == infeedRow {
            
            if infeedView != nil {
                infeedView!.removeFromSuperview()
                let cell:UITableViewCell =  UITableViewCell.init(style: .default, reuseIdentifier: "cell")
                cell.contentView.clipsToBounds  = true
                cell.contentView.addSubview(infeedView!)
                cell.selectionStyle = .none
                return cell
            }
            
        }
        
        
        // create a new cell if needed or reuse an old one
//        let cell:UITableViewCell = self.tableView.dequeueReusableCell(withIdentifier: cellReuseIdentifier) as! UITableViewCell
        let cell:UITableViewCell =  UITableViewCell.init(style: .default, reuseIdentifier: "cell")
        // set the text from the data model
        cell.textLabel?.text = "Test Infeed"
        
        return cell
    }
    
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        if indexPath.row == infeedRow {
            if infeedView != nil {
                return self.view.frame.size.width * (3/5)
            }else {
                return 0
            }
            
        }else {
            return 60
        }
    }
    
//    func tableView(_ tableView: UITableView, estimatedHeightForRowAt indexPath: IndexPath) -> CGFloat {
//        if indexPath.row == infeedRow {
//            if infeedView != nil {
//                return self.view.frame.size.width * (3/5)
//            }else {
//                return 0
//            }
//
//        }else {
//            return 60
//        }
//    }
    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */
    @IBAction func openMenu(_ sender: Any) {
        APP_DELEGATE.drawerViewController?.openMenu()
    }
    
    func dealloc()  {
        infeedFactory?.releaseMemory()
        infeedView = nil
    }
}
